massa-execution-worker/src/execution.rs

```diff
@@ -769,50 +831,33 @@ impl ExecutionState {
         Ok(context_guard!(self).settle_slot())
     }
 
-    /// Gets a full ledger entry both at the latest final and active executed slots
-    /// TODO: this can be heavily optimized, see comments and `https://github.com/massalabs/massa/issues/2343`
-    ///
-    /// # returns
-    /// `(final_entry, active_entry)`
-    pub fn get_final_and_active_ledger_entry(
+    /// Gets a parallel balance both at the latest final and active executed slots
+    pub fn get_final_and_active_parallel_balance(
         &self,
-        addr: &Address,
-    ) -> (Option<LedgerEntry>, Option<LedgerEntry>) {
-        // get the full entry from the final ledger
-        let final_entry = self.final_state.read().ledger.get_full_entry(addr);
-
-        // get cumulative active changes and apply them
-        // TODO there is a lot of overhead here: we only need to compute the changes for one entry and no need to clone it
-        // also we should proceed backwards through history for performance
-        // https://github.com/massalabs/massa/issues/2343
-        // Note that get_accumulated_active_changes_at_slot is called at the slot AFTER the active one
-        // in order to take all available active slots into account (and not forget the last one)
-        // and prevent a get_accumulated_active_changes_at_slot crash in the case active_cursor = final_cursor.
+        address: &Address,
+    ) -> (Option<Amount>, Option<Amount>) {
+        let final_balance = self.final_state.read().ledger.get_parallel_balance(address);
         let next_slot = self
             .active_cursor
             .get_next_slot(self.config.thread_count)
             .expect("slot overflow when getting speculative ledger");
-        let active_change = self
-            .get_accumulated_active_changes_at_slot(next_slot)
-            .ledger_changes
-            .get(addr)
-            .cloned();
-        let active_entry = match (&final_entry, active_change) {
-            (final_v, None) => final_v.clone(),
-            (_, Some(SetUpdateOrDelete::Set(v))) => Some(v),
-            (_, Some(SetUpdateOrDelete::Delete)) => None,
-            (None, Some(SetUpdateOrDelete::Update(u))) => {
-                let mut v = LedgerEntry::default();
-                v.apply(u);
-                Some(v)
-            }
-            (Some(final_v), Some(SetUpdateOrDelete::Update(u))) => {
-                let mut v = final_v.clone();
-                v.apply(u);
-                Some(v)
-            }
-        };
+        let active_balance = self.get_active_balance_at_slot(next_slot, address);
```

Comment:
Note that if the balance is not found in the history, this should do a 2nd lookup into the final_state.
This has a performance impact and should be avoided because we already did this lookup above